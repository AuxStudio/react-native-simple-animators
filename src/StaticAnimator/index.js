import React from 'react';
import PropTypes from 'prop-types';
import { Animated, Easing } from 'react-native';

import utils from '../utils';

export default class StaticAnimator extends React.Component {
  constructor(props) {
    super(props);

    this.reset = this.reset.bind(this);
    this.animateIn = this.animateIn.bind(this);
    this.animateOut = this.animateOut.bind(this);

    this.state = {
      animatedValue: new Animated.Value(0),
    };
  }

  static get propTypes() {
    return {
      type: PropTypes.string.isRequired, // any react-native style prop
      initialValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      finalValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      shouldAnimateIn: PropTypes.bool,
      shouldAnimateOut: PropTypes.bool,
      animateInCallBack: PropTypes.func,
      animateInCallback: PropTypes.func,
      animateOutCallBack: PropTypes.func,
      animateOutCallback: PropTypes.func,
      shouldRepeat: PropTypes.bool,
      shouldLoop: PropTypes.bool, // used with repeat (cycles the animation instead of going back and forth)
      style: PropTypes.oneOfType([PropTypes.number, PropTypes.object, PropTypes.array]),
      duration: PropTypes.number,
      easing: PropTypes.func,
      delay: PropTypes.number,
      children: PropTypes.node,
    };
  }

  static get defaultProps() {
    return {
      duration: 250,
      easing: Easing.gentle,
    };
  }

  componentDidMount() {
    const shouldAnimateOnMount =
      (this.props.initialValue || this.props.initialValue === 0) &&
      (this.props.finalValue || this.props.finalValue === 0);

    if (this.props.shouldAnimateIn && shouldAnimateOnMount) {
      this.animateIn();
    }
  }

  componentDidUpdate(prevProps) {
    const shouldReset = this.props.type && this.props.type !== prevProps.type;

    const shouldAnimateOnUpdate =
      (this.props.initialValue && this.props.initialValue !== prevProps.initialValue) ||
      (this.props.finalValue && this.props.finalValue !== prevProps.finalValue) ||
      shouldReset;

    if (shouldReset) {
      this.reset(shouldAnimateOnUpdate);
    } else if (shouldAnimateOnUpdate) {
      this.animateIn();
    } else if (this.props.shouldAnimateIn && !prevProps.shouldAnimateIn) {
      this.animateIn();
    } else if (this.props.shouldAnimateOut && !prevProps.shouldAnimateOut) {
      this.animateOut();
    }
  }

  reset(shouldAnimateIn) {
    this.setState(
      {
        animatedValue: new Animated.Value(0),
      },
      () => {
        if (shouldAnimateIn) {
          this.animateIn();
        }
      },
    );
  }

  animateIn() {
    const useNativeDriver = utils.isTransform(this.props.type) || this.props.type === 'opacity';

    Animated.timing(this.state.animatedValue, {
      toValue: 1,
      duration: this.props.duration,
      easing: this.props.easing,
      useNativeDriver,
      delay: this.props.delay,
    }).start(() => {
      if (this.props.animateInCallBack) {
        this.props.animateInCallBack();
      } else if (this.props.animateInCallback) {
        this.props.animateInCallback();
      }

      if (this.props.shouldRepeat) {
        if (this.props.shouldLoop) {
          this.state.animatedValue.setValue(0);
          this.animateIn();
        } else {
          this.animateOut();
        }
      }
    });
  }

  animateOut() {
    const useNativeDriver = utils.isTransform(this.props.type) || this.props.type === 'opacity';

    Animated.timing(this.state.animatedValue, {
      toValue: 0,
      duration: this.props.duration,
      easing: this.props.easing,
      useNativeDriver,
      delay: this.props.delay,
    }).start(() => {
      if (this.props.animateOutCallBack) {
        this.props.animateOutCallBack();
      } else if (this.props.animateOutCallback) {
        this.props.animateOutCallback();
      }

      if (this.props.shouldRepeat) {
        this.animateIn();
      }
    });
  }

  render() {
    // Append deg to rotate animations
    const initialValue =
      this.props.type === 'rotate' ? `${this.props.initialValue}deg` : this.props.initialValue;
    const finalValue =
      this.props.type === 'rotate' ? `${this.props.finalValue}deg` : this.props.finalValue;

    const animatedTypeStyle = {};
    animatedTypeStyle[this.props.type] = this.state.animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: [initialValue, finalValue],
    });

    let animatedStyles;

    // If the type of animation is a transform, push the animatedTypeStyle to an array of 'transform'
    if (utils.isTransform(this.props.type)) {
      animatedStyles = {
        transform: [animatedTypeStyle],
      };
    } else {
      animatedStyles = animatedTypeStyle;
    }

    return (
      <Animated.View style={[this.props.style, animatedStyles]}>
        {this.props.children}
      </Animated.View>
    );
  }
}
