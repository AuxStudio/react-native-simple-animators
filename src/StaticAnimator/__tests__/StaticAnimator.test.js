import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';

import StaticAnimator from '../';

// FIXME: shouldAnimateOut, delay, duration, shouldRepeat and shouldLoop were occluded from these tests
// due to the async complexity they introduce
describe('StaticAnimator', () => {
  let spy;

  it('renders with a height animation to 100', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="height"
          initialValue={0}
          finalValue={100}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with a height animation to 50', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="height"
          initialValue={0}
          finalValue={50}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with a height animation to 0', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="height"
          initialValue={0}
          finalValue={0}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with a height animation to 0', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="height"
          initialValue={0}
          finalValue={0}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with a translateX animation to 100', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="translateX"
          initialValue={0}
          finalValue={100}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with an opacity animation to 1', () => {
    expect(
      renderer.create(
        <StaticAnimator
          type="height"
          initialValue={0}
          finalValue={1}
          shouldAnimateIn
          animateInCallback={jest.fn()}
          easing={jest.fn()}
        >
          <View style={{ width: 100, height: 100 }} />
        </StaticAnimator>,
      ),
    ).toMatchSnapshot();
  });

  it('animateIn is not called on componentDidMount', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'animateIn');

    renderer.create(<StaticAnimator type="height" initialValue={0} finalValue={1} />);

    expect(spy).not.toHaveBeenCalled();
  });

  it('animateIn is called on componentDidMount', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'animateIn');

    renderer.create(
      <StaticAnimator type="height" initialValue={0} finalValue={1} shouldAnimateIn />,
    );

    expect(spy).toHaveBeenCalled();
  });

  it('animateIn is called on componentDidUpdate', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'animateIn');

    const component = renderer.create(
      <StaticAnimator type="height" initialValue={0} finalValue={1} />,
    );

    expect(spy).not.toHaveBeenCalled();

    component.update(
      <StaticAnimator type="height" initialValue={0} finalValue={1} shouldAnimateIn />,
    );

    expect(spy).toHaveBeenCalled();
  });

  it('reset is called on componentDidUpdate when the type prop changes', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'reset');

    const component = renderer.create(
      <StaticAnimator type="height" initialValue={0} finalValue={1} />,
    );

    expect(spy).not.toHaveBeenCalled();

    component.update(<StaticAnimator type="width" initialValue={0} finalValue={1} />);

    expect(spy).toHaveBeenCalled();
  });

  it('animateIn is called on componentDidUpdate when the values change', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'animateIn');

    const component = renderer.create(
      <StaticAnimator type="height" initialValue={0} finalValue={1} shouldAnimateIn />,
    );

    expect(spy).toHaveBeenCalled();

    component.update(
      <StaticAnimator type="height" initialValue={1} finalValue={0} shouldAnimateIn />,
    );

    expect(spy).toHaveBeenCalled();
  });

  it('animateOut is called on componentDidUpdate', () => {
    spy = jest.spyOn(StaticAnimator.prototype, 'animateOut');

    const component = renderer.create(
      <StaticAnimator type="height" initialValue={0} finalValue={1} />,
    );

    component.update(
      <StaticAnimator type="height" initialValue={0} finalValue={1} shouldAnimateOut />,
    );

    expect(spy).toHaveBeenCalled();
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
  });
});
