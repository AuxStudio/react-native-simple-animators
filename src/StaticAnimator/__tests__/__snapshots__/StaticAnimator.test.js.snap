// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a StaticAnimator with a height animation to 0 1`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`renders a StaticAnimator with a height animation to 0 2`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`renders a StaticAnimator with a height animation to 50 1`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`renders a StaticAnimator with a height animation to 100 1`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`renders a StaticAnimator with a translateX animation to 100 1`] = `
<View
  style={
    Object {
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`renders a StaticAnimator with an opacity animation to 1 1`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <View
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;
